boot/freeldr/freeldr/arch/i386/pc/machpc.c

```diff
@@ -177,6 +177,63 @@ PcGetHarddiskConfigurationData(UCHAR DriveNumber, ULONG* pSize)
     return PartialResourceList;
 }
 
+static
+VOID
+DetectDockingStation(PCONFIGURATION_COMPONENT_DATA BusKey)
+{
+    PCM_PARTIAL_RESOURCE_LIST PartialResourceList;
+    PCM_PARTIAL_RESOURCE_DESCRIPTOR PartialDescriptor;
+    PCONFIGURATION_COMPONENT_DATA ControllerKey;
+    ULONG Size, Result;
+    PDOCKING_STATE_INFORMATION DockingState;
+
+    Result = PnpBiosGetDockStationInformation(DiskReadBuffer);
+
+    /* Build full device descriptor */
+    Size = sizeof(CM_PARTIAL_RESOURCE_LIST) + sizeof(DOCKING_STATE_INFORMATION);
+
+    if (Result == 0)
+    {
+        /* FIXME: Add more device specific data */
+        ERR("FIXME: System docked\n");
+    }
+
+    PartialResourceList = FrLdrHeapAlloc(Size, TAG_HW_RESOURCE_LIST);
+    if (PartialResourceList == NULL)
+    {
+        ERR("Failed to allocate resource descriptor\n");
+        return;
+    }
+
+    /* Initialize resource descriptor */
+    RtlZeroMemory(PartialResourceList, Size);
+    PartialResourceList->Version = 0;
+    PartialResourceList->Revision = 0;
+    PartialResourceList->Count = 1;
+
+    /* Set device specific data */
+    PartialDescriptor = &PartialResourceList->PartialDescriptors[0];
+    PartialDescriptor->Type = CmResourceTypeDeviceSpecific;
+    PartialDescriptor->ShareDisposition = CmResourceShareUndetermined;
+    PartialDescriptor->Flags = 0;
+    PartialDescriptor->u.DeviceSpecificData.DataSize = sizeof(DOCKING_STATE_INFORMATION);
+
+    DockingState = (PDOCKING_STATE_INFORMATION)&PartialResourceList->PartialDescriptors[1];
+    DockingState->ReturnCode = Result;
+
+    /* Create controller key */
+    FldrCreateComponentKey(BusKey,
+                           AdapterClass, // FIXME: Is this correct?
+                           DockingInformation,
+                           0,
+                           0,
+                           0xFFFFFFFF,
+                           "Docking State Information",
+                           PartialResourceList,
+                           Size,
+                           &ControllerKey);
```

Comment:
In view of the previous comment, it's certainly not a controller key you are creating, but a component one.